lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -349,6 +357,107 @@ def _promtail_error(self, event):
+
+Let's say that we have a charm's workload that writes logs to the standard output (stdout),
+and we need to send those logs to a workload implementing the `loki_push_api` interface,
+such as `Loki` or `Grafana Agent`. To know how to reach a Loki instance, a charm would
+typically use the `loki_push_api` interface.
+
+Use the `LogForwarder` class by instantiating it in the `__init__` method of the charm:
+
+```python
+from charms.loki_k8s.v1.loki_push_api import LogForwarder
+
+...
+
+  def __init__(self, *args):
+      ...
+      self._log_forwarder = LogForwarder(
+          self,
+          relation_name="logging"  # optional, defaults to `logging`
+      )
+```
+
+The `LogForwarder` by default will observe relation events on the `logging` endpoint and
+enable/disable log forwarding automatically.
+Next, modify the `metadata.yaml` file to add:
+
+The `log-forwarding` relation in the `requires` section:
+```yaml
+requires:
+ logging:
+   interface: loki_push_api
+   optional: true
+```
+
+Once the library is implemented in a charm and a logging relation (loki_push_api) is
+active and healthy, the library will inject a Pebble layer in the workload container
+to configure Pebble's log forwarding feature and start sending logs to Loki.
+
+## ManualLogForwarder class Usage
+
+Similarly to the `LogForwarder` class, we have a charm's workload that writes logs to the
+standard output (stdout), and we need to send those logs to some Loki-compatible endpoints.
+
+However, if you charm doesn't directly relate to another charm implementing the `loki_push_api`
+interface, this class allows you to explicitly specify the `loki_endpoints` in the constructor.
+
+To handle updating the log forwarding layers, you have two options:
+
+1. Manually call `update_logging()` from your charm whenever you want the log forwarding layer
+   to be updated.
+
+   ```python
+   ...
+
+       def __init__(self, *args):
+           ...
+           self._log_forwarder = ManualLogForwarder(
+                self,
+                loki_endpoints=get_endpoints_from_somewhere()
+            )
+
+           self.framework.observe(self.on.some_event, self._log_forwarder.update_logging())
+
+       def get_endpoints_from_somewhere():
+           # Example function that gets the endpoints from Juju config.
+           #
+           # Returns:
+           #     A dictionary of remote units and respective Loki endpoint.
+           #     {
+           #         "loki/0": "http://loki1:3100/loki/api/v1/push",
+           #     }
+           #
+           return self.config["loki_endpoints"]
+   ```
+
+2. Alternatively, you can pass a `relation_name` to the constructor; `update_logging` will
+   automatically be called whenever that relation triggers an event:
+
+   ```python
+   ...
+
+       def __init__(self, *args):
+           ...
+           self._log_forwarder = ManualLogForwarder(
+                self,
+                loki_endpoints=get_endpoints_from_somewhere(),
```

Comment:
```suggestion
                loki_endpoints=self.get_endpoints_from_somewhere(),
```

is that a callback, or a function call?

Reply:
Should be a function call if I understand correctly.